### ISSUE TITLE
Requesting review

### ISSUE THREAD

**Issue by mciantyre** [imxrt-rs/imxrt-boot-gen#1]
Here's the `imxrt1062-fcb-gen` from the `teensy4-rs` project. It's more advanced than the original crate, and it's gone through a rename. I'm looking for some early feedback before we start considering it.

How's the name? I went with `boot-gen` because it generates data structures required for booting. I changed it from `fcb-gen`, which had meant "firmware configuration block" (FCB). I was abusing that term. The FCB is a specific data structure in the serial NAND boot process. I confused the abbreviation with FlexSPI configuration block (FCB), which is general to all serial FLASH boot data structures. But, if we stick with `fcb-gen`, we can't extend the crate to include SEMC (parallel FLASH) booting! `boot-gen` is all I could think of. (I also want to play with IVT and DCD hooks in this crate, and the generic name leaves us open to that possibility.)

How's the feature-flag utilization? We require a feature flag, just as we require a feature flag in the RAL crate. The feature flags don't do much, since we only support the most common feature: an FCB for serial NOR FLASH over FlexSPI.

Will it actually work with the iMXRT1011 EVK board? I studied the 1010 reference manual, and compared it to the 1060 reference manual and implementation. I found a few discrepancies, and I accounted for them behind feature flags. From my understanding, this crate should be ready to generate a 1011's FCB for serial NOR flash boot.

Finally, and maybe most importantly, is this all too much? Seems like a lot of overhead to generate a byte array. I justify it by saying the approach is self-documenting, and it could help defend against silly programming errors. It's the opposite of the Teensy approach, which [hard-codes the array](https://github.com/PaulStoffregen/cores/blob/10084eca06ebfffc70d37d628a3a3a18d92fc498/teensy4/bootdata.c#L39). Let me know if this all seems too complex for what it's actually doing.

**Comment by bfrog** [comment]
First I think this is a nice way of doing this. Looking at how the NXP SDK does it with some compiler specific attributes on a C struct, which may-or-may not pan out depending on how you configure it, I think this method provides a very easy method of doing things and can be checked at compile time using the builder interface. Bravo. NXP themselves call it a boot header I believe in their SDK, this being a generator for that, the name makes sense.

I don't think its that much overhead, its part of the chip itself. One thing to keep in mind is that in the NXP SDK the boot header is disabled when running from ram presumably for debugging, not sure if that's something that's easy to account for, probably easy enough to document and use a application specific feature flag for perhaps.

**Comment by bfrog** [comment]
I think it might be helpful to specify how much of this

https://community.nxp.com/docs/DOC-341317

This project tackles in the readme, and what the goals are.

All of it?
I don't know, it would be nice to have digital signing of images too at some point far in the future.

**Comment by mciantyre** [comment]
Thanks for the thoughts! We'll stick with this approach, unless we feel a good reason to change. Like you mentioned, it's should give us a good path forward for adding other capabilities.

I've made some changes to the API. I realized that the approach I was proposing wouldn't be backwards-compatible. If we kept with the current pattern, we would end up making more public fields, and users would be required to add those fields, else suffer breaking changes. I've changed the API to [follow a real builder](https://rust-lang.github.io/api-guidelines/type-safety.html#c-builder). Let me know if the previous API was favorable.

Also thanks for the guide link. I'll keep NXP SDK conventions in mind as we move forward.

**Comment by mciantyre** [comment]
I was also considering an approach that resembled

```rust
use imxrt_boot::serial_flash::*;

#[link_section = ".whatever.you.want"]
#[no_mangle]
pub static YOUR_FCB_NAME: FCB = fcb! {
    read_sample_clock_source: ReadSampleClockSource::LoopbackFromDQSPad,
    cs_hold_time: CSHoldTime(0x01),
    cs_setup_time: CSSetupTime(0x02),
    column_address_width: ColumnAddressWidth::OtherDevices,
    device_mode_cfg_enabled: 0,
    wait_time_cfg_commands: 0,
    s_flash_a1_size: SerialFlashSize::new(0x0020_0000),
    // Other fields...
};
```

The library exposes the `FCB` type, which looks like

```rust
#[repr(C)]
pub struct FCB {
    #[doc(hidden)]
    pub _tag: u32,
    #[doc(hidden)]
    pub _version: u32,
    #[doc(hidden)]
    pub _reserved0: [u8; 4],
    pub read_sample_clock_source: ReadSampleClockSource,
    pub cs_hold_time: CSHoldTime,
    pub cs_setup_time: CSSetupTime,
    pub column_address_width: ColumnAddressWidth,
    pub device_mode_cfg_enabled: u8,
    #[doc(hidden)]
    pub _reserved1: [u8; 1],
    // Other fields...
}
```

The `fcb!()` macro would handle the initialization of the hidden fields, and provide a compile-time interface for defining the FCB. The `FCB` memory is laid out just as we need it. This is the "how it might be done in C" approach. This skips the whole `build.rs` process to auto-generate code. Users are responsible for defining the symbol name and link section.

I didn't finish prototyping this, so I'm not sure if it would work or scale. Just putting it out there in case anyone wants to consider it!
